### PR TITLE
Make RESTMapper in client cluster-aware

### PR DESCRIPTION
**Pull request by sttts**
This PR makes discovery cluster-aware, i.e. the RESTMapper is instantiated per cluster and cached with an LRU cache.

This matters because currently the controller-runtime client is unusable on non-uniform virtual workspaces or even against kcp itself as a scoped client because the used REST mapper talks to the wildcard endpoint only.

### CODE REVIEW

File: pkg/kcp/wrappers.go [hunk 1]
```diff
@@ -248,3 +272,47 @@ func generatePath(originalPath string, clusterPath logicalcluster.Path) string {
...
+func cachedRESTMapper(config *rest.Config) (func(ctx context.Context) (meta.RESTMapper, error), error) {
+	httpClient, err := rest.HTTPClientFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var mu sync.RWMutex
+	cache := expirable.NewLRU[logicalcluster.Path, meta.RESTMapper](1000, nil, time.Minute)
+	return func(ctx context.Context) (meta.RESTMapper, error) {
+		c, _ := kontext.ClusterFrom(ctx)
```
Comment (soumyadip007) [inline pkg/kcp/wrappers.go]: The error from `kontext.ClusterFrom(ctx)` is ignored `(c, _ := kontext.ClusterFrom(ctx))`. We may consider it.

```
c, err := kontext.ClusterFrom(ctx)
if err != nil {
    // Handle error appropriately, perhaps returning a nil mapper with the error
    return nil, err
}
```

**Review by soumyadip007 — APPROVED** [review]

File: pkg/kcp/wrappers.go (hunk 1 above)
Comment (sttts) [reply pkg/kcp/wrappers.go]: Noticed that the second return value is a boolean, not an error. Will add a comment.

Comment (sttts) [reply pkg/kcp/wrappers.go]: done

File: pkg/client/client.go [hunk 2]
```diff
@@ -197,11 +210,16 @@ func newClient(config *rest.Config, options Options) (*client, error) {
 		},
 		metadataClient: metadataClient{
 			client:     rawMetaClient,
-			restMapper: options.Mapper,
+			restMapper: options.MapperWithContext,
 		},
 		scheme: options.Scheme,
 		mapper: options.Mapper,
 	}
+	mapperCache, err := lru.New[logicalcluster.Name, meta.RESTMapper](1000)
```
Comment (mjudeikis) [inline pkg/client/client.go]: Would it make sense 1000 to create a global variable? If somebody wants to tune it, its possible this way?

**Review by mjudeikis — APPROVED** [review]
just nit hypotatical question. else - lgtm to me. Will wait for @embik cycle

File: pkg/client/client.go (hunk 2 above)
Comment (sttts) [reply pkg/client/client.go]: Addressed.

**Pull request merged**